controllers/auth.js

```diff
@@ -9,6 +9,99 @@ const {
   USER_DOES_NOT_EXIST_ERROR,
 } = require("../constants/errorMessages");
 
+const googleAuthLogin = (req, res, next) => {
+  const { redirectURL, dev } = req.query;
+  if (dev === "true") {
+    return passport.authenticate("google", {
+      scope: ["email"],
+      state: redirectURL,
+    })(req, res, next);
+  } else {
+    return res.boom.unauthorized("User cannot be authenticated");
+  }
+};
+
+async function handleGoogleLogin(req, res, user, authRedirectionUrl) {
+  const rdsUiUrl = new URL(config.get("services.rdsUi.baseUrl"));
+  try {
+    if (!user.emails || user.emails.length === 0) {
+      logger.error("Google login failed: No emails found in user data");
+      return res.boom.unauthorized("No email found in Google account");
+    }
+    const primaryEmail = user.emails.find((email) => email.verified === true);
+    if (!primaryEmail) {
+      logger.error("Google login failed: No verified email found");
+      return res.boom.unauthorized("No verified email found in Google account");
+    }
+
+    const userData = {
+      email: primaryEmail.value,
+      created_at: Date.now(),
+      updated_at: null,
+    };
+
+    const userDataFromDB = await users.fetchUser({ email: userData.email });
+
+    if (userDataFromDB.userExists) {
+      if (userDataFromDB.user.roles?.developer) {
+        const errorMessage = encodeURIComponent("Google login is restricted for developer role.");
+        const separator = authRedirectionUrl.search ? "&" : "?";
+        return res.redirect(`${authRedirectionUrl}${separator}error=${errorMessage}`);
+      }
+    }
```

Comment:
Can you please explain this part?

Reply:
I’m checking if the user exists in the database, and if they do, I then verify whether they have a "developer" role. As Ankush told me, if the user has the developer role, we're preventing them from logging in via Google.

The user is then redirected back to the original URL with an error query parameter, which contains a message indicating that Google login is restricted for users with the "developer" role. This error parameter will be used on the frontend to display a toast notification to inform the user of the restriction.

Reply:
so basically you are passing an error message in the query param? is this the correct approach?

Reply:
To clarify, passing error messages via query parameters is consistent with how OAuth providers like Google handle errors. When authentication fails, Google (and others) typically pass error codes and descriptions through the URL, as explained [here](https://www.oauth.com/oauth2-servers/server-side-apps/possible-errors/).

Since our backend handles the OAuth flow and user verification, we need to send error feedback to the frontend. Given that the OAuth redirect directly returns to the backend, and the frontend can't interact with the OAuth process in real-time, using query parameters is the most efficient way to pass error messages to the frontend for display.

Please let me know if you believe this approach is not correct.